File: appengine/src/py/handlers/rhizo.py
#!/usr/bin/env python
#
# Copyright 2010 The Rhizosphere Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import os.path

from google.appengine.ext import webapp
from google.appengine.ext.webapp import template
from google.appengine.ext.webapp.util import run_wsgi_app

from py import rhizoglobals

class RhizoHandler(webapp.RequestHandler):

    def _getOptionFromUrl(self, optname, allowedvalues):
      opt = self.request.get(optname)
      if opt and opt not in allowedvalues:
        return allowedvalues[0]
      return opt

    def get(self):
        # Startup options
        platform, device = rhizoglobals.IdentifyPlatformDevice(
            self.request)
        uitemplate = self.request.get('template')

        # Experimental features
        experimental = rhizoglobals.GetOptionFromUrl(self.request,
                                                     'exp', ['0', '1'])
        if experimental == '1':
          use_channels = True
        else:
          use_channels = False

        template_values = rhizoglobals.DefaultTemplate(self.request)
        template_values.update({
            'template': uitemplate,
            'device': device,
            'platform': platform,
            'use_channels': use_channels,
        })
        path = os.path.join(os.path.dirname(__file__),
                            '../../templates%s' % self.request.path)
        self.response.out.write(template.render(path, template_values))


class IGoogleHandler(webapp.RequestHandler):

    def get(self):
        template_values = rhizoglobals.DefaultTemplate(self.request)
        template_values.update({
            'hostname': 'http://%s' % os.environ['HTTP_HOST']
        })
        path = os.path.join(os.path.dirname(__file__), '../../templates/ig.xml')
        self.response.headers.add_header('Content-Type', 'text/xml')
        self.response.out.write(template.render(path, template_values))


application = webapp.WSGIApplication(
    [('/rhizo.html', RhizoHandler),
     ('/multi.html', RhizoHandler),
     ('/google_visualization.html', RhizoHandler),
     ('/ig', IGoogleHandler),
    ],
    debug=rhizoglobals.appenginedebug)


def main():
    run_wsgi_app(application)


if __name__ == '__main__':
    main()

File: src/js/rhizo.js
/*
  Copyright 2008 The Rhizosphere Authors. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

function namespace(ns) {
  var nsParts = ns.split(".");
  var root = window;

  for (var i=0; i<nsParts.length; i++) {
    if (typeof root[nsParts[i]] == "undefined") {
      root[nsParts[i]] = {};
    }
    root = root[nsParts[i]];
  }
}

namespace("rhizo");

/**
 * Inherit the prototype methods from one constructor into another.
 */
rhizo.inherits = function(childCtor, parentCtor) {
  function tempCtor() {};
  tempCtor.prototype = parentCtor.prototype;
  childCtor.prototype = new tempCtor();
  childCtor.prototype.constructor = childCtor;
};

namespace("rhizo.util");

rhizo.util.LOG_E_10 = Math.log(10);

/**
   Returns the log_10() of a number.
 */
rhizo.util.log10_ = function(val) {
  return Math.log(val)/rhizo.util.LOG_E_10;
};

/**
   Returns the order of magnitude of a given number.
   E.g. for a number between 100K and 1M returns 5.
   if 0 is passed in, NaN is returned.
   @param {number} num the number to evaluate.
 */
rhizo.util.orderOfMagnitude = function(num) {
  if (num == 0) {
    return Number.NaN;
  }

  // The multiply/divide by 100 trick is just to get rid of rounding errors.
  // For example, in Safari/MacOs log10(1M) would result in 5.999 instead of 6
  // and consequently floored to 5. And we cannot remove floor().
  return Math.floor(Math.round(rhizo.util.log10_(Math.abs(num))*100)/100);
};

/**
 * Parses an URI extracting the different parts that compose it.
 *
 * This function comes straight from Steven Levithan's parseURI library
 * (http://blog.stevenlevithan.com/archives/parseuri), released under the MIT
 * license. See the NOTICE file for further info.
 * @param {string} str The URI to parse.
 * @return {Object.<string, string>} A key-value map for all the composing
 *     parts of the parsed URI, like 'protocol', 'host', 'path' and so forth.
 */
rhizo.util.parseUri = function(str) {
  var o = rhizo.util.parseUri.options_;
  var m = o.parser[o.strictMode ? "strict" : "loose"].exec(str);
  var uri = {};
  var i   = 14;

  while (i--) uri[o.key[i]] = m[i] || "";

  uri[o.q.name] = {};
  uri[o.key[12]].replace(o.q.parser, function ($0, $1, $2) {
    if ($1) uri[o.q.name][$1] = $2;
  });

  return uri;
};

/**
 * Configuration options for rhizo.util.parseUri
 * @type {*}
 * @private
 */
rhizo.util.parseUri.options_ = {
  strictMode: false,
  key: ["source","protocol","authority","userInfo","user","password","host",
        "port","relative","path","directory","file","query","anchor"],
  q: {
    name:   "queryKey",
    parser: /(?:^|&)([^&=]*)=?([^&]*)/g
  },
  parser: {
    strict: /^(?:([^:\/?#]+):)?(?:\/\/((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?))?((((?:[^?#\/]*\/)*)([^?#]*))(?:\?([^#]*))?(?:#(.*))?)/,
    loose:  /^(?:(?![^:@]+:[^:@\/]*@)([^:\/?#.]+):)?(?:\/\/)?((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?)(((\/(?:[^?#](?![^?#\/]*\.[^?#\/.]+(?:[?#]|$)))*\/?)?([^?#\/]*))(?:\?([^#]*))?(?:#(.*))?)/
  }
};

/**
 * @return {Object.<stirng, string>} A key-value map of all the URL parameters
 *     in the current document URL.
 */
rhizo.util.urlParams = function() {
  return rhizo.util.parseUri(document.location.href).queryKey;
};